Rule "curly" with "multi" has false positive on nested if-for-if, fixer changes code behavior

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are here for another reason, please see below:

    1. To propose a new rule: https://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a rule change: https://eslint.org/docs/developer-guide/contributing/rule-changes
    3. To request a change that is not a bug fix, rule change, or new rule: https://eslint.org/docs/developer-guide/contributing/changes
    4. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

<!--
    If you are using ESLint v6.5.0 or later, you can run ESLint with the `--env-info` flag and paste the output here.
-->

* **ESLint Version:** 6.8.0
* **Node Version:** 12.16.0
* **npm Version:** 6.13.4

This bug also reproduces on ESLint 7.0.0-alpha.1.

**What parser (default, Babel-ESLint, etc.) are you using?**
Default parser

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```js
{
  "rules": {
    "curly": ["warn", "multi"]
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

<!-- Paste the source code below: -->
```js
var x = [''];
if (x)
  for (var i in x) {
    if (i > 0)
      console.log(i);
  }
else
  console.log('whoops');
```

<!-- Paste the command you used to run ESLint: -->
```bash
./node_modules/.bin/eslint --fix index.js
```

**What did you expect to happen?**
ESLint should make no changes to the code, as the only curly braces present are required.

**What actually happened? Please include the actual, raw output from ESLint.**
ESLint changes the code to the following:

```js
var x = [''];
if (x)
  for (var i in x) 
    if (i > 0)
      console.log(i);
  
else
  console.log('whoops');
```

This is incorrect -- the `else` is now associated with the inner `if` instead of the outer one. The execution behavior changes accordingly:

```shell
$ node index.js 
$ ./node_modules/.bin/eslint index.js

/tmp/eslint-repro/index.js
  3:3  warning  Unnecessary { after 'for-in'  curly

✖ 1 problem (0 errors, 1 warning)
  0 errors and 1 warning potentially fixable with the `--fix` option.

$ ./node_modules/.bin/eslint --fix index.js 
$ node index.js 
whoops
```

I did a little bit more experimentation: when I replace the `for` block with a `while` block, the same issue occurs. However, when I replace the `for` block with another `if`, ESLint correctly detects that the curly braces are mandatory.

**Are you willing to submit a pull request to fix this bug?**
I can if necessary, although the fix is probably much faster for somebody who's already familiar with the relevant code.

--- a/lib/rules/curly.js
+++ b/lib/rules/curly.js
@@ -141,33 +141,14 @@ module.exports = {
         }
 
         /**
-         * Checks a given IfStatement node requires braces of the consequent chunk.
-         * This returns `true` when below:
-         *
-         * 1. The given node has the `alternate` node.
-         * 2. There is a `IfStatement` which doesn't have `alternate` node in the
-         *    trailing statement chain of the `consequent` node.
-         * @param {ASTNode} node A IfStatement node to check.
-         * @returns {boolean} `true` if the node requires braces of the consequent chunk.
+         * Determines whether the given node has an `else` keyword token as the first token after.
+         * @param {ASTNode} node The node to check.
+         * @returns {boolean} `true` if the node is followed by an `else` keyword token.
          */
-        function requiresBraceOfConsequent(node) {
-            if (node.alternate && node.consequent.type === "BlockStatement") {
-                if (node.consequent.body.length >= 2) {
-                    return true;
-                }
+        function isFollowedByElseKeyword(node) {
+            const nextToken = sourceCode.getTokenAfter(node);
 
-                for (
-                    let currentNode = node.consequent.body[0];
-                    currentNode;
-                    currentNode = astUtils.getTrailingStatement(currentNode)
-                ) {
-                    if (currentNode.type === "IfStatement" && !currentNode.alternate) {
-                        return true;
-                    }
-                }
-            }
-
-            return false;
+            return Boolean(nextToken) && isElseKeywordToken(nextToken);
         }
 
         /**
@@ -225,6 +206,110 @@ module.exports = {
         }
 
         /**
+         * Determines whether the code represented by the given node contains an `if` statement
+         * that would become associated with an `else` keyword directly appended to that code.
+         *
+         * Examples where it returns `true`:
+         *
+         *    if (a)
+         *        foo();
+         *
+         *    if (a) {
+         *        foo();
+         *    }
+         *
+         *    if (a)
+         *        foo();
+         *    else if (b)
+         *        bar();
+         *
+         *    while (a)
+         *        if (b)
+         *            if(c)
+         *                foo();
+         *            else
+         *                bar();
+         *
+         * Examples where it returns `false`:
+         *
+         *    if (a)
+         *        foo();
+         *    else
+         *        bar();
+         *
+         *    while (a) {
+         *        if (b)
+         *            if(c)
+         *                foo();
+         *            else
+         *                bar();
+         *    }
+         *
+         *    while (a)
+         *        if (b) {
+         *            if(c)
+         *                foo();
+         *        }
+         *        else
+         *            bar();
+         * @param {ASTNode} node Node representing the code to check.
+         * @returns {boolean} `true` if an `if` statement within the code would become associated with an `else` appended to that code.
+         */
+        function hasUnsafeIf(node) {
+            switch (node.type) {
+                case "IfStatement":
+                    if (!node.alternate) {
+                        return true;
+                    }
+                    return hasUnsafeIf(node.alternate);
+                case "ForStatement":
+                case "ForInStatement":
+                case "ForOfStatement":
+                case "LabeledStatement":
+                case "WithStatement":
+                case "WhileStatement":
+                    return hasUnsafeIf(node.body);
+                default:
+                    return false;
+            }
+        }
+
+        /**
+         * Determines whether the existing curly braces around the single statement are necessary to preserve the semantics of the code.
+         * The braces, which make the given block body, are necessary in either of the following situations:
+         *
+         * 1. The statement is a lexical declaration.
+         * 2. Without the braces, an `if` within the statement would become associated with an `else` after the closing brace:
+         *
+         *     if (a) {
+         *         if (b)
+         *             foo();
+         *     }
+         *     else
+         *         bar();
+         *
+         *     if (a)
+         *         while (b)
+         *             while (c) {
+         *                 while (d)
+         *                     if (e)
+         *                         while(f)
+         *                             foo();
+         *            }
+         *     else
+         *         bar();
+         * @param {ASTNode} node `BlockStatement` body with exactly one statement directly inside. The statement can have its own nested statements.
+         * @returns {boolean} `true` if the braces are necessary - removing them (replacing the given `BlockStatement` body with its single statement content)
+         * would change the semantics of the code or produce a syntax error.
+         */
+        function areBracesNecessary(node) {
+            const statement = node.body[0];
+
+            return isLexicalDeclaration(statement) ||
+                hasUnsafeIf(statement) && isFollowedByElseKeyword(node);
+        }
+
+        /**
          * Prepares to check the body of a node to see if it's a block statement.
          * @param {ASTNode} node The node to report if there's a problem.
          * @param {ASTNode} body The body node to check for blocks.
@@ -242,30 +327,29 @@ module.exports = {
             const hasBlock = (body.type === "BlockStatement");
             let expected = null;
 
-            if (node.type === "IfStatement" && node.consequent === body && requiresBraceOfConsequent(node)) {
+            if (hasBlock && (body.body.length !== 1 || areBracesNecessary(body))) {
                 expected = true;
             } else if (multiOnly) {
-                if (hasBlock && body.body.length === 1 && !isLexicalDeclaration(body.body[0])) {
-                    expected = false;
-                }
+                expected = false;
             } else if (multiLine) {
                 if (!isCollapsedOneLiner(body)) {
                     expected = true;
                 }
-            } else if (multiOrNest) {
-                if (hasBlock && body.body.length === 1 && isOneLiner(body.body[0])) {
-                    const leadingComments = sourceCode.getCommentsBefore(body.body[0]);
-                    const isLexDef = isLexicalDeclaration(body.body[0]);
 
-                    if (isLexDef) {
-                        expected = true;
-                    } else {
-                        expected = leadingComments.length > 0;
-                    }
-                } else if (!isOneLiner(body)) {
-                    expected = true;
+                // otherwise, the body is allowed to have braces or not to have braces
+
+            } else if (multiOrNest) {
+                if (hasBlock) {
+                    const statement = body.body[0];
+                    const leadingCommentsInBlock = sourceCode.getCommentsBefore(statement);
+
+                    expected = !isOneLiner(statement) || leadingCommentsInBlock.length > 0;
+                } else {
+                    expected = !isOneLiner(body);
                 }
             } else {
+
+                // default "all"
                 expected = true;
             }
 

--- a/tests/lib/rules/curly.js
+++ b/tests/lib/rules/curly.js
@@ -336,6 +336,89 @@ ruleTester.run("curly", rule, {
             // https://github.com/feross/standard/issues/664
             code: "if (true) foo()\n;[1, 2, 3].bar()",
             options: ["multi-line"]
+        },
+
+        // https://github.com/eslint/eslint/issues/12928 (also in invalid[])
+        {
+            code: "if (x) for (var i in x) { if (i > 0) console.log(i); } else console.log('whoops');",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) { if (b) foo(); } else bar();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) { if (b) foo(); } else bar();",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (a) { if (b) foo(); } else { bar(); }",
+            options: ["multi", "consistent"]
+        },
+        {
+            code: "if (a) { if (b) foo(); } else { bar(); }",
+            options: ["multi-or-nest", "consistent"]
+        },
+        {
+            code: "if (a) { if (b) { foo(); bar(); } } else baz();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) foo(); else if (b) { if (c) bar(); } else baz();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) { if (b) foo(); else if (c) bar(); } else baz();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) if (b) foo(); else { if (c) bar(); } else baz();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) { lbl:if (b) foo(); } else bar();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) { lbl1:lbl2:if (b) foo(); } else bar();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) { for (;;) if (b) foo(); } else bar();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) { for (key in obj) if (b) foo(); } else bar();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) { for (elem of arr) if (b) foo(); } else bar();",
+            options: ["multi"],
+            parserOptions: { ecmaVersion: 2015 }
+        },
+        {
+            code: "if (a) { with (obj) if (b) foo(); } else bar();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) { while (cond) if (b) foo(); } else bar();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) { while (cond) for (;;) for (key in obj) if (b) foo(); } else bar();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) while (cond) { for (;;) for (key in obj) if (b) foo(); } else bar();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) while (cond) for (;;) { for (key in obj) if (b) foo(); } else bar();",
+            options: ["multi"]
+        },
+        {
+            code: "if (a) while (cond) for (;;) for (key in obj) { if (b) foo(); } else bar();",
+            options: ["multi"]
         }
     ],
     invalid: [
@@ -1091,6 +1174,114 @@ ruleTester.run("curly", rule, {
             output: "do \ndoSomething()\n;\n while (foo)",
             options: ["multi-or-nest"],
             errors: [{ messageId: "unexpectedCurlyAfter", data: { name: "do" }, type: "DoWhileStatement" }]
+        },
+
+        // https://github.com/eslint/eslint/issues/12928 (also in valid[])
+        {
+            code: "if (a) { if (b) foo(); }",
+            output: "if (a)  if (b) foo(); ",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if (a) { if (b) foo(); else bar(); }",
+            output: "if (a)  if (b) foo(); else bar(); ",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if (a) { if (b) foo(); else bar(); } baz();",
+            output: "if (a)  if (b) foo(); else bar();  baz();",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if (a) { while (cond) if (b) foo(); }",
+            output: "if (a)  while (cond) if (b) foo(); ",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if (a) while (cond) { if (b) foo(); }",
+            output: "if (a) while (cond)  if (b) foo(); ",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "while" }, type: "WhileStatement" }]
+        },
+        {
+            code: "if (a) while (cond) { if (b) foo(); else bar(); }",
+            output: "if (a) while (cond)  if (b) foo(); else bar(); ",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "while" }, type: "WhileStatement" }]
+        },
+        {
+            code: "if (a) { while (cond) { if (b) foo(); } bar(); baz() } else quux();",
+            output: "if (a) { while (cond)  if (b) foo();  bar(); baz() } else quux();",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "while" }, type: "WhileStatement" }]
+        },
+        {
+            code: "if (a) { if (b) foo(); } bar();",
+            output: "if (a)  if (b) foo();  bar();",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if(a) { if (b) foo(); } if (c) bar(); else baz();",
+            output: "if(a)  if (b) foo();  if (c) bar(); else baz();",
+            options: ["multi-or-nest"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if (a) { do if (b) foo(); while (cond); } else bar();",
+            output: "if (a)  do if (b) foo(); while (cond);  else bar();",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if (a) do { if (b) foo(); } while (cond); else bar();",
+            output: "if (a) do  if (b) foo();  while (cond); else bar();",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfter", data: { name: "do" }, type: "DoWhileStatement" }]
+        },
+        {
+            code: "if (a) { if (b) foo(); else bar(); } else baz();",
+            output: "if (a)  if (b) foo(); else bar();  else baz();",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if (a) while (cond) { bar(); } else baz();",
+            output: "if (a) while (cond)  bar();  else baz();",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "while" }, type: "WhileStatement" }]
+        },
+        {
+            code: "if (a) { for (;;); } else bar();",
+            output: "if (a)  for (;;);  else bar();",
+            options: ["multi"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if (a) { while (cond) if (b) foo() } else bar();",
+            output: "if (a) { while (cond) if (b) foo() } else {bar();}",
+            options: ["multi", "consistent"],
+            errors: [{ messageId: "missingCurlyAfter", data: { name: "else" }, type: "IfStatement" }]
+        },
+        {
+
+            /**
+             * Reports 2 errors, but one pair of braces is necessary if the other pair gets removed.
+             * Auto-fix will remove only outer braces in the first iteration.
+             * After that, the inner braces will become valid and won't be removed in the second iteration.
+             * If user manually removes inner braces first, the outer braces will become valid.
+             */
+            code: "if (a) { while (cond) { if (b) foo(); } } else bar();",
+            output: "if (a)  while (cond) { if (b) foo(); }  else bar();",
+            options: ["multi"],
+            errors: [
+                { messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" },
+                { messageId: "unexpectedCurlyAfterCondition", data: { name: "while" }, type: "WhileStatement" }
+            ]
         }
     ]
 });

